tests/conftest.py

```diff
@@ -340,6 +344,24 @@ def app_config(database):
     return cfg
 
 
+@pytest.fixture(scope="session")
+def app_config(database):
+
+    return get_app_config(
+        database,
+    )
+
+
+@pytest.fixture(scope="session")
+def app_config_dbsession_from_env(database):
```

Comment:
Seeing this split out is helpful - since it allows me to ask the following question:

Historically, in the context of test scenarios we've been passing a session-scoped `app_config` fixture that has a `database` fixture passed in.

Is there any reason for us to maintain these two fixtures, or could we collapse these down to `app_config` only?

Reply:
The `database` is current a session-wide fixture like `app_config`. I suppose that since `app_config` depends on the other fixture, then `database` is created firstly.

However, trying to include the code from `database` in `app_config` gave birth to this error:

```
_________________________________________________ ERROR at teardown of test_user_profile _________________________________________________

    @request.addfinalizer
    def drop_database():
>       janitor.drop()

tests/conftest.py:336: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../lib/python3.11/site-packages/pytest_postgresql/janitor.py:90: in drop
    self._dont_datallowconn(cur, db_to_drop)
../lib/python3.11/site-packages/pytest_postgresql/janitor.py:98: in _dont_datallowconn
    cur.execute(f'ALTER DATABASE "{dbname}" with allow_connections false;')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <psycopg.Cursor [closed] [BAD] at 0x7ce3261f5c70>, query = 'ALTER DATABASE "tests" with allow_connections false;', params = None

    def execute(
        self,
        query: Query,
        params: Optional[Params] = None,
        *,
        prepare: Optional[bool] = None,
        binary: Optional[bool] = None,
    ) -> Self:
        """
        Execute a query or command to the database.
        """
        try:
            with self._conn.lock:
                self._conn.wait(
                    self._execute_gen(query, params, prepare=prepare, binary=binary)
                )
        except e._NO_TRACEBACK as ex:
>           raise ex.with_traceback(None)
E           psycopg.errors.InvalidCatalogName: database "tests" does not exist

../lib/python3.11/site-packages/psycopg/cursor.py:732: InvalidCatalogName
```

I'm not sure, but I think what's happening is that since `request.addfinalizer` adds something to a finalization stack, then _maybe_ some garbage collection has destroyed something (i.e. the database) already at the point where `janitor.drop()` is supposed to be called.

I tried out some different sequences, but in the end I think it's maybe easier for now to keep the old fixtures structure.

Reply:
So actually this error was very helpful to understand that a bit more refactoring is needed to arrive at the right place :+1: 

Will try to figure out a minimal change :hourglass_flowing_sand:

Reply:
> Is there any reason for us to maintain these two fixtures, or could we collapse these down to app_config only?

So, getting back to this question, I can think of 2 reasons to keep them separate:

1) To keep McCabe complexity low :innocent: Not sure if you have mccabe rules on or off in flake8, but very long routines get hard to read and maintain. There's a clear responsibility defined for each of these two fixtures, so seems easy to maintain separate fixtures.

2) Actually pytest can help here: It knows how to reuse the same instance of a fixture, even if the path is different (see diagram below). So while the `app_config` fixture may have a variant, the `database` fixture remains the same.

I'm exploring this alternative approach: https://github.com/benjaoming/warehouse/compare/issue_15634_test_user_profile-investigation...benjaoming:warehouse:issue_15634_test_user_profile-investigation-refactor-dbsession?expand=1

1) It avoids pulling in `db_session` in `webtest` but instead creates the session, using its own variant of `app_config` (`app_config_dbsession_from_env`).

2) ...which effectively avoids having 2 app configs instantiated when only 1 is needed in the session.

A quick drawing (trying out SmartDraw, hence a bit messy)

![image](https://github.com/pypi/warehouse/assets/374612/00081933-4cde-4de2-bd47-5a18b14788ce)

Seeing this does make me want to add some documentation w ascii/mermaid diagrams to visualize how fixtures depend on each other.

There's also something about the names of fixtures and helper functions that could be more intuitive? `webtest` is actually Warehouse's own "test client" - that is, if we look at Django or FastAPI, the name of the fixture would be better as `client` or simply `app` (Pyramid terminology).

<details><summary>Similar conftest.py for FastAPI</summary>
<p>

```
import pytest

from fastapi.testclient import TestClient
from main import create_app


@pytest.fixture
def app():
    return create_app()


@pytest.fixture
def client(app):
    return TestClient(app)
```

</p>
</details>